File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function getDifference(arr, n){
    // splice한 wires로 tree 만들기
    const tree = new Map();
    arr.forEach(([a, b]) =>{
        tree.has(a) ? tree.get(a).push(b) : tree.set(a, [b]);
        tree.has(b) ? tree.get(b).push(a) : tree.set(b, [a]);
    })
    // 순회
    const checked = [];
    const queue = [];
    queue.push(arr[0][0]);
    
    while(queue.length){
        const node = queue.pop();
        if(!checked.includes(node)){
            checked.push(node);
            queue.push(...tree.get(node));
        }
    }
    // 순회길이를 이용해 차이 구하기
    // console.log(checked)
    return Math.abs(checked.length- (n - checked.length))
}

function solution(n, wires) {
    const difference = wires.map((el, idx) => {
        const temp = [...wires];
        temp.splice(idx, 1);
        return getDifference(temp, n)
    })
    return Math.min(...difference)
}

File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def get_difference(arr, n):
    # tree 만들기
    tree = {}
    for a, b in arr:
        if a in tree:
            tree[a].append(b);
        else:
            tree[a] = [b];
        if b in tree:
            tree[b].append(a);
        else:
            tree[b] = [a];
    
    # 순회
    checked = [];
    queue = [];
    queue.append(arr[0][0]);
    while queue:
        node = queue.pop();
        if node not in checked:
            checked.append(node);
            queue.extend(tree.get(node, [])) ;
    # 순회 길이를 이용히 차이 구하기
    return abs(len(checked) - (n - len(checked)))

def solution(n, wires):
    # wires 를 돌면서 간선 하나씩을 삭제하고 getDifference(차이를 구하는 함수)에 넣어줌
    difference = [get_difference(wires[:i] + wires[i+1:], n) for i in range(len(wires))];
    # 차이가 가장 적은 값 return 
    return min(difference)
